### PR TITLE
KAFKA-5819: Add Joined class and relevant KStream join overloads

### PR DESCRIPTION
**Pull request by dguy**
Add the `Joined` class and the overloads to `KStream` that use it.
Deprecate existing methods that have `Serde` params

### CODE REVIEW

**Comment by dguy** [comment]
@guozhangwang, @mjsax, @bbejeck

File: streams/src/main/java/org/apache/kafka/streams/kstream/Joined.java [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.streams.kstream;
+
+import org.apache.kafka.common.serialization.Serde;
+
+/**
+ * The Joined class represents Optional params that can be passed to
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/Joined.java]: `{@code Joined}`
`[o]ptional para[meter]s`

File: streams/src/main/java/org/apache/kafka/streams/kstream/Joined.java [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+ * {@link KStream#join}, {@link KStream#leftJoin}, and  {@link KStream#outerJoin} operations
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/Joined.java]: `.`

File: streams/src/main/java/org/apache/kafka/streams/kstream/Joined.java [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+ */
+public class Joined<K, V, VO> {
+
+    private Serde<K> keySerde;
+    private Serde<V> valueSerde;
+    private Serde<VO> otherValueSerde;
+
+    private Joined(final Serde<K> keySerde, final Serde<V> valueSerde, final Serde<VO> otherValueSerde) {
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/Joined.java]: nit: parameter/line formatting

File: streams/src/main/java/org/apache/kafka/streams/kstream/Joined.java [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+     *
+     * @param keySerde        the key serde to use. If null the default key serde from config will be used
+     * @param valueSerde      the value serde to use. If null the default value serde from config will be used
+     * @param otherValueSerde the otherValue serde to use. If null the default value serde from config will be used
+     * @param <K>             key type
+     * @param <V>             value type
+     * @param <VO>            other value type
+     * @return new {@code Joined} instance with the provided serdes
+     */
+    public static <K, V, VO> Joined<K, V, VO> with(final Serde<K> keySerde, final Serde<V> valueSerde, final Serde<VO> otherValueSerde) {
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/Joined.java]: formatting

File: streams/src/main/java/org/apache/kafka/streams/kstream/Joined.java [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+    private Serde<VO> otherValueSerde;
+
+    private Joined(final Serde<K> keySerde, final Serde<V> valueSerde, final Serde<VO> otherValueSerde) {
+        this.keySerde = keySerde;
+        this.valueSerde = valueSerde;
+        this.otherValueSerde = otherValueSerde;
+    }
+
+    /**
+     * Create an instance of {@code Joined} with key, value, and otherValue {@link Serde} instances. Null values are
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/Joined.java]: line/sentence formatting
`{@code null}`.

File: streams/src/main/java/org/apache/kafka/streams/kstream/Joined.java [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+     * accepted and will be replaced by the default serdes as defined in config
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/Joined.java]: .

File: streams/src/main/java/org/apache/kafka/streams/kstream/Joined.java [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+     *
+     * @param keySerde        the key serde to use. If null the default key serde from config will be used
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/Joined.java]: `{@code null}` -- same below

File: streams/src/main/java/org/apache/kafka/streams/kstream/Joined.java [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+     * @param <VO>            other value type
+     * @return new {@code Joined} instance with the provided serdes
+     */
+    public static <K, V, VO> Joined<K, V, VO> with(final Serde<K> keySerde, final Serde<V> valueSerde, final Serde<VO> otherValueSerde) {
+        return new Joined<>(keySerde, valueSerde, otherValueSerde);
+    }
+
+    /**
+     * Create an instance of {@code Joined} with  a key {@link Serde}. Null values are
+     * accepted and will be replaced by the default key serde as defined in config
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/Joined.java]: as above

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java [hunk 9]
```diff
@@ -619,39 +634,58 @@ public boolean test(final K1 key, final V1 value) {
...
-        return join(other, joiner, null, null);
+        return join(other, joiner, Joined.<K, V, V1>with(null, null, null));
     }
 
     @Override
-    public <V1, R> KStream<K, R> join(final KTable<K, V1> other,
-                                      final ValueJoiner<? super V, ? super V1, ? extends R> joiner,
-                                      final Serde<K> keySerde,
-                                      final Serde<V> valueSerde) {
+    public <VT, VR> KStream<K, VR> join(final KTable<K, VT> other, final ValueJoiner<? super V, ? super VT, ? extends VR> joiner, final Joined<K, V, VT> joined) {
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java]: parameter/line formatting

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamImplTest.java [hunk 10]
```diff
@@ -373,4 +375,27 @@ public void shouldNotAllowNullJoinerOnLeftJoinWithGlobalTable() throws Exception
                         null);
     }
 
+    @Test(expected = NullPointerException.class)
+    public void shouldThrowNullPointerOnLeftJoinWithTableWhenJoinedIsNull() {
+        testStream.leftJoin(builder.table(Serdes.String(), Serdes.String(), "blah"),
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamImplTest.java]: Should weextract `builder.table` to make sure there will be no NPE there ?

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamJoinTest.java [hunk 11]
```diff
@@ -170,7 +174,9 @@ public void testOuterJoin() throws Exception {
         processor = new MockProcessorSupplier<>();
         stream1 = builder.stream(intSerde, stringSerde, topic1);
         stream2 = builder.stream(intSerde, stringSerde, topic2);
-        joined = stream1.outerJoin(stream2, MockValueJoiner.TOSTRING_JOINER, JoinWindows.of(100), intSerde, stringSerde, stringSerde);
+        joined = stream1.outerJoin(stream2, MockValueJoiner.TOSTRING_JOINER,
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamJoinTest.java]: new line for `MockValueJoiner`

**Comment by guozhangwang** [comment]
retes this please

**Comment by guozhangwang** [comment]
retest this please

**Comment by guozhangwang** [comment]
LGTM provided jenkins success.

**Comment by dguy** [comment]
merged to trunk

**Pull request closed without merging**